chain/actors/actor_storagemarket.go

```diff
@@ -0,0 +1,541 @@
+package actors
+
+import (
+	"bytes"
+
+	"github.com/filecoin-project/go-amt-ipld"
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-hamt-ipld"
+	cbg "github.com/whyrusleeping/cbor-gen"
+
+	"github.com/filecoin-project/go-lotus/build"
+	"github.com/filecoin-project/go-lotus/chain/actors/aerrors"
+	"github.com/filecoin-project/go-lotus/chain/address"
+	"github.com/filecoin-project/go-lotus/chain/types"
+)
+
+type StorageMarketActor struct{}
+
+type smaMethods struct {
+	Constructor                  uint64
+	WithdrawBalance              uint64
+	AddBalance                   uint64
+	CheckLockedBalance           uint64
+	PublishStorageDeals          uint64
+	HandleCronAction             uint64
+	SettleExpiredDeals           uint64
+	ProcessStorageDealsPayment   uint64
+	SlashStorageDealCollateral   uint64
+	GetLastExpirationFromDealIDs uint64
+	ActivateStorageDeals         uint64
+}
+
+var SMAMethods = smaMethods{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11}
+
+func (sma StorageMarketActor) Exports() []interface{} {
+	return []interface{}{
+		2: sma.WithdrawBalance,
+		3: sma.AddBalance,
+		// 4: sma.CheckLockedBalance,
+		5: sma.PublishStorageDeals,
+		// 6: sma.HandleCronAction,
+		// 7: sma.SettleExpiredDeals,
+		// 8: sma.ProcessStorageDealsPayment,
+		// 9: sma.SlashStorageDealCollateral,
+		// 10: sma.GetLastExpirationFromDealIDs,
+		11: sma.ActivateStorageDeals, // TODO: move under PublishStorageDeals after specs team approves
+	}
+}
+
+type StorageParticipantBalance struct {
+	Locked    types.BigInt
+	Available types.BigInt
+}
+
+type StorageMarketState struct {
+	Balances cid.Cid // hamt<addr, StorageParticipantBalance>
+	Deals    cid.Cid // amt<StorageDeal>
+
+	NextDealID uint64 // TODO: amt.LastIndex()
+}
+
+type StorageDealProposal struct {
+	PieceRef           cid.Cid // can this mess anything up? Should this just be cid bytes
+	PieceSize          uint64
+	Client             address.Address
+	Provider           address.Address
+	ProposalExpiration uint64
+	DealExpiration     uint64
+	StoragePrice       types.BigInt
+	StorageCollateral  types.BigInt
+	ProposerSignature  types.Signature
+}
+
+type StorageDeal struct {
+	Proposal         StorageDealProposal
+	CounterSignature types.Signature
+}
+
+type OnChainDeal struct {
+	Deal            StorageDeal
+	ActivationEpoch uint64 // 0 = inactive
+}
+
+type WithdrawBalanceParams struct {
+	Balance types.BigInt
+}
+
+func (sma StorageMarketActor) WithdrawBalance(act *types.Actor, vmctx types.VMContext, params *WithdrawBalanceParams) ([]byte, ActorError) {
+	var self StorageMarketState
+	old := vmctx.Storage().GetHead()
+	if err := vmctx.Storage().Get(old, &self); err != nil {
+		return nil, err
+	}
+
+	b, bnd, err := getMarketBalances(vmctx, self.Balances, []address.Address{vmctx.Message().From})
+	if err != nil {
+		return nil, aerrors.Wrap(err, "could not get balance")
+	}
+
+	balance := b[0]
```

Comment:
Probably want to do a sanity check on the length here

Reply:
This length is always going to be equal to length of the passed address array. If it's not this should just panic